fontconfig: Stop using versioned config dirs

jtojnar opened the pull request:
###### Motivation for this change
Turns out lot of software (including Chromium) use bundled fontconfig so we either need to wrap every one of those, or re-introduce the global unversioned config. The latter is easier but weakens hermetic configs.

But perhaps those are not really worth the effort. Given that programs linked against fc 2.12.6 on fc 2.13.92 system seem to at least display text, even while printing tons of errors (as long as you generate fc cache manually), and same thing the other way around, hopefully it will not be an issue in the future. If we give up on the hermetic configs, we can get rid of lot of complexity (see the diff).

###### Links

* fontconfig MR improving the severity of unrelated compat warnings: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/merge_requests/111
* https://logs.nix.samueldr.com/nixos-dev/2020-08-13#1597332919-1597339347;

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jtojnar committed:
Revert "fontconfig: Only read versioned config dirs"

This reverts commit edf2541f02c6b24ea791710d5cadeae36f9b1a3a.

Turns out lot of software (including Chromium) use bundled fontconfig
so we either need to wrap every one of those, or re-introduce the global
unversioned config. The latter is easier but weakens hermetic configs.

But perhaps those are not really worth the effort. Given that programs
linked against fc 2.12.6 on fc 2.13.92 system seem to at least display text,
even while printing tons of errors (as long as you generate fc cache manually),
and same thing the other way around, hopefully it will not be an issue in the future.

We give up on the hermetic configs in exchange for getting rid of lot of complexity.

jtojnar committed:
fontconfig: get rid of rest of versioned configs

The incompatibility does not seem to exist any more: programs linked against fc 2.12
on fc 2.14 system seem to at least display text, even while printing tons of errors
(as long as you generate fc cache manually), and same thing the other way around.
Hopefully it will not be an issue in the future.

jtojnar commented:
Let’s fix the chromium in master first.

flokli commented:
https://github.com/NixOS/nixpkgs/pull/95362 was merged, so master should be fixed.

jtojnar commented:
I am completely convinced this is the way to go.

jtojnar commented on pkgs/development/libraries/fontconfig/default.nix:
```diff
@@ -106,32 +94,22 @@ stdenv.mkDerivation rec {
+    # Don't try to write to /var/cache/fontconfig at install time.
+    "fc_cachedir=$(TMPDIR)/dummy"
+    "RUN_FC_CACHE_TEST=false"
+    "sysconfdir=${placeholder "out"}/etc"
```
Yes, the configure makes `/etc` hardcoded into the code but we cannot install to `/etc` so we need to move the installation path into the derivation.

flokli replied:
Okay :+1:

flokli commented:
We might need to merge https://github.com/NixOS/nixpkgs/pull/95716 first (and then rebase this PR on top of master)

flokli commented:
@jtojnar can this be rebased?

jtojnar commented:
Rebased.

worldofpeace commented:
@jtojnar I think we should document the fontconfig changes in the release notes eventually.

edolstra commented:
Not convinced this is a good idea. What are we going to do when fonfconfig changes in a backwards-incompatible way again in the future?

jtojnar commented:
The versioned config was introduced in https://github.com/NixOS/nixpkgs/commit/0927405a371292e9abddb2365540e46e2f1cc2c7 and https://github.com/NixOS/nixpkgs/commit/c0e2aceef46033e43ab8b4f6aa0b36b25ba35f3d (https://github.com/NixOS/nixpkgs/pull/4410) to resolve https://github.com/NixOS/nixpkgs/issues/2050, there were also some issues with `xml:space` https://github.com/NixOS/nixpkgs/commit/b16994f7ce49ac1793837879f8e3ddc63c976a5f

Looking at the changes in https://github.com/NixOS/nixpkgs/pull/4410#issuecomment-58927000 it was really just minor relaxation of the schema. The unknown elements would already be skipped with a warning and the patch I sent upstream changes the unknown attributes severity to warning as well (and they were already ignored previously). For that reason backwards compatibility should not be an issue – the configs will continue to be parsed unless they change the format drastically, which is pretty unlikely with the upstream development pace. Forwards compatibility should not be an issue for the same reason.

Outside of NixOS, we would use the unversioned config in either case so nothing changes there.

jtojnar commented:
The versioned config was introduced in https://github.com/NixOS/nixpkgs/commit/0927405a371292e9abddb2365540e46e2f1cc2c7 and https://github.com/NixOS/nixpkgs/commit/c0e2aceef46033e43ab8b4f6aa0b36b25ba35f3d (https://github.com/NixOS/nixpkgs/pull/4410) to resolve https://github.com/NixOS/nixpkgs/issues/2050, there were also some issues with `xml:space` https://github.com/NixOS/nixpkgs/commit/b16994f7ce49ac1793837879f8e3ddc63c976a5f

Looking at the changes in https://github.com/NixOS/nixpkgs/pull/4410#issuecomment-58927000 it was really just minor relaxation of the schema. The unknown elements would already be skipped with a warning and the patch I sent upstream changes the unknown attributes severity to warning as well (and they were already ignored previously). For that reason backwards compatibility should not be an issue – the configs will continue to be parsed unless they change the format drastically, which is pretty unlikely with the upstream development pace. Forwards compatibility should not be an issue for the same reason.

Outside of NixOS, we would use the unversioned config in either case so nothing changes there.

jtojnar commented:
The only reason that comes to mind why we might still want to have versioning is to generate cache for multiple fontconfig versions declaratively but we are not really supporting multiple versions any way. And we could do that in a single config and fontconfig should ignore the cache files generated by other versions.

jtojnar committed:
makeFontsConf: clean up

The use of the argument was removed in 2016 but its definition was left behind.

https://github.com/NixOS/nixpkgs/commit/cd2948a72e3116a3a42f1f523fe2638db994362c

jtojnar commented:
Merging after IRC discussion with RM.

vcunat commented:
I wanted to try some tests on latest `staging-next`
```sh
nix build -f nixos/release-combined.nix nixos.tests.installer.swraid.x86_64-linux
```
but got blocked by
```
builder for '/nix/store/HASH1-fontconfig-conf.drv' failed with exit code 1; last 1 log lines:
  ln: failed to create symbolic link '/nix/store/HASH2-fontconfig-conf/etc/fonts/conf.d/50-user.conf': File exists
```

_You'll probably know way faster than me._

flokli commented on pkgs/development/libraries/fontconfig/default.nix:
```diff
@@ -106,32 +94,22 @@ stdenv.mkDerivation rec {
       > fonts.conf.tmp
     mv fonts.conf.tmp $out/etc/fonts/fonts.conf
-
-    # Make it easier to remove user config in NixOS module.
```
@vcunat this probably should have been kept?

flokli replied:
(or https://github.com/NixOS/nixpkgs/pull/95358/files#diff-376672cba12c87d4103e6134632e306fR203 remove the symlink in case `cfg.includeUserConf` is false?

jtojnar replied:
I forgot to revert also 8425726f86a2f4a38d0022f3d5cb1d2001da6999.

jtojnar commented:
Should be fixed by f0cb5c6a150bea1b8b7fc9dc3b0cdecceb0bdc54.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-builds-a-ton-of-packages-instead-of-fetching-them-in-the-cache/13231/8
